ext/uri/php_uri_common.h

```diff
@@ -0,0 +1,190 @@
+*/
+
+#ifndef PHP_URI_COMMON_H
+#define PHP_URI_COMMON_H
+
+extern zend_class_entry *whatwg_url_ce;
+extern zend_object_handlers whatwg_uri_object_handlers;
+extern zend_class_entry *uri_comparison_mode_ce;
+extern zend_class_entry *uri_exception_ce;
+extern zend_class_entry *invalid_uri_exception_ce;
+extern zend_class_entry *whatwg_invalid_url_exception_ce;
+extern zend_class_entry *whatwg_url_validation_error_type_ce;
+extern zend_class_entry *whatwg_url_validation_error_ce;
+
+typedef enum {
+	URI_RECOMPOSITION_RAW_ASCII,
+	URI_RECOMPOSITION_RAW_UNICODE,
+	URI_RECOMPOSITION_NORMALIZED_ASCII,
+	URI_RECOMPOSITION_NORMALIZED_UNICODE,
+} uri_recomposition_mode_t;
+
+typedef enum {
+	URI_COMPONENT_READ_RAW,
+	URI_COMPONENT_READ_NORMALIZED_ASCII,
+	URI_COMPONENT_READ_NORMALIZED_UNICODE,
+} uri_component_read_mode_t;
+
+typedef struct uri_handler_t {
+	const char *name;
+
+	zend_result (*init_parser)(void);
+	void *(*parse_uri)(const zend_string *uri_str, const void *base_url, zval *errors);
+	void (*create_invalid_uri_exception)(zval *exception_zv, zval *errors);
+	void *(*clone_uri)(void *uri);
+	zend_string *(*uri_to_string)(void *uri, uri_recomposition_mode_t recomposition_mode, bool exclude_fragment);
+	void (*free_uri)(void *uri);
+	zend_result (*destroy_parser)(void);
+	HashTable *property_handlers;
+} uri_handler_t;
+
+typedef struct uri_internal_t {
+	const uri_handler_t *handler;
+	void *uri;
+} uri_internal_t;
+
+typedef struct uri_object_t {
+	uri_internal_t internal;
+	zend_object std;
+} uri_object_t;
+
+static inline uri_object_t *uri_object_from_obj(const zend_object *object) {
+	return (uri_object_t*)((char*)(object) - XtOffsetOf(uri_object_t, std));
+}
+
+static inline uri_internal_t *uri_internal_from_obj(const zend_object *object) {
+	return &(uri_object_from_obj(object)->internal);
+}
+
+#define Z_URI_OBJECT_P(zv) uri_object_from_obj(Z_OBJ_P((zv)))
+#define Z_URI_INTERNAL_P(zv) uri_internal_from_obj(Z_OBJ_P((zv)))
+
+#define URI_PARSER_WHATWG "Uri\\WhatWg\\Url"
+#define URI_SERIALIZED_PROPERTY_NAME "uri"
+
+typedef zend_result (*uri_read_t)(const uri_internal_t *internal_uri, uri_component_read_mode_t read_mode, zval *retval);
+
+typedef zend_result (*uri_write_t)(uri_internal_t *internal_uri, zval *value, zval *errors);
+
+typedef struct uri_property_handler_t {
+	uri_read_t read_func;
+	uri_write_t write_func;
+} uri_property_handler_t;
+
+#define URI_REGISTER_PROPERTY_READ_HANDLER(property_handlers, name, property_read_func) do { \
+	static const uri_property_handler_t handler = {.read_func = property_read_func, .write_func = NULL}; \
+	uri_register_property_handler(property_handlers, name, &handler); \
+} while (0)
+
+#define URI_REGISTER_PROPERTY_READ_WRITE_HANDLER(property_handlers, name, property_read_func, property_write_func) do { \
+	static const uri_property_handler_t handler = {.read_func = property_read_func, .write_func = property_write_func}; \
```

Comment:
It took me a while to see why this isn't just storing a stack memory value (I missed the static). I think we can do without the macro by using `zend_hash_add_new_mem()` in `uri_register_property_handler()` and then just calling:

```
	uri_register_property_handler(&lexbor_property_handlers, ZSTR_KNOWN(ZEND_STR_SCHEME), (uri_property_handler_t){
		.read_func = lexbor_read_scheme,
		.write_func = lexbor_write_scheme,
	});
```

like this. It's a little more verbose, but very clear in what is happening.

Reply:
Actually, the current approach was taken over from the DOM extension. DOM used to have the _add_mem approach you describe, but I stepped away from it because the data doesn't need to be on the heap and can sit in rodata

Reply:
I see. I'll have another look at this, when I perform an "in-depth" review after the first round of fixes.